[evals] Add support for scaling evals with ray

erictang000 opened the pull request:
# What does this PR do?
This PR adds support for using ray to speed up evals and data generation. Currently we use are using a preliminary version of using ray data + vllm while we wait for the code at [ray.llm](https://github.com/ray-project/ray/tree/master/python/ray/llm) to be fully open sourced (coming in the next 1-2 weeks), after which we will migrate over.

## How to Use
To use the new path
- first add the submodule by running `git submodule update --init`
- then simply add `--use_ray` to existing commands and set relevant scaling parameters in `--ray_config`. 

Reasonable defaults and examples of how to set advanced vllm engine arguments are provided in `ray_configs/ray_config.yaml`. For example, to run the Math-500 eval with `Sky-T1-32B-Preview`, you can use the following command 
```bash
python inference_and_check.py --model NovaSky-AI/Sky-T1-32B-Preview --dataset MATH500  --split test --max_tokens 8192 --use_ray --ray_config ray_configs/ray_config.yaml
```
where the `ray_config` looks like:
```yaml
llm_engine: vllm # currently only vllm supported
accelerator_type: A100-80G # accelerator name as specified here: https://docs.ray.io/en/master/ray-core/accelerator-types.html#accelerator-types
engine_kwargs: # vllm engine kwargs 
  tensor_parallel_size: 2
  gpu_memory_utilization: 0.9
runtime_env:
  env_vars:
    VLLM_ATTENTION_BACKEND: "FLASH_ATTN"
env_config:
  num_replicas: 4 # number of vllm replicas 
  batch_size: 128 # ray data internal batch size (used for map_batches call internally). Should usually be set to a value in [64, 128, 256] for best performance.
```

## Speedups
<img width="1004" alt="image" src="https://github.com/user-attachments/assets/140f5e31-869b-4b97-bbdc-e2378c823751" />
Example speedups using `ray` can be seen above - we found that for a single 8xA100 or 8xH100 node, setting `tp=4` with 2 replicas or `tp=2` with 4 replicas using ray can be much faster than a single `tp=8` replica, especially for longer evals like MMLU.

erictang000 committed:
add rayllm batch path

erictang000 committed:
fix typo

erictang000 committed:
temp make mmlu pro smaller

erictang000 committed:
update vllm version, rayllm config, add repartition

erictang000 committed:
Add submodule repo

erictang000 committed:
move evalworkload code to outside module

erictang000 committed:
Update submodule to latest commit

erictang000 committed:
remove [:4000] for mmlupro

erictang000 committed:
updates to inference_and_save path

erictang000 committed:
fix small issues

erictang000 committed:
disable n > 1 for inference and save rayllm path

erictang000 committed:
Merge branch 'rayllm' of https://github.com/erictang000/SkyThought into rayllm

erictang000 committed:
inference_and_save works for n = 1 use_rayllm

erictang000 committed:
Remove submodule

erictang000 committed:
remove submodule and rename to pipeline

erictang000 committed:
remove unnecessary model_id

erictang000 committed:
remove .gitmodules

erictang000 committed:
rename main to pipeline

erictang000 committed:
add support for n > 1

erictang000 committed:
remove old code

erictang000 committed:
fix unflatten logic for n > 1

erictang000 committed:
merge

SumanthRH commented on skythought/tools/inference_and_check.py:
Has this been tested with multi node yet? 

Would be nice to do so.

SumanthRH replied:
Mainly the cpu head node + multiple GPU worker nodes pattern

erictang000 replied:
yes i've tested a bunch multi-node on the L40S: See https://docs.google.com/spreadsheets/d/12NDF6ttGPHeyx8CKkkG3OPnzGgV5aax4uORMtQkqEh4/edit?usp=sharing

erictang000 replied:
all with that pattern

erictang000 committed:
finish merge stuff

erictang000 committed:
split small datasets

SumanthRH commented on skythought/skythought_evals/inference_and_check.py:
```diff
@@ -505,10 +566,16 @@ def main():
     args = parser.parse_args()
     set_seed(args.seed)
 
+    # use os to enable hf_transfer for model download
+    os.environ["HF_HUB_ENABLE_HF_TRANSFER"] = "1"
```
1. Can you add `hf_transfer` to requirements.txt?
2. This needs a guard. If `HF_HUB_ENABLE_HF_TRANSFER` is already set by the user, we don't do anything. If it's unset, we set it to 1. (to allow user overrides)

SumanthRH commented on skythought/skythought_evals/ray_configs/ray_config.yaml:
```diff
@@ -0,0 +1,23 @@
+    VLLM_ATTENTION_BACKEND: "FLASH_ATTN"
+env_config:
+  num_replicas: 2 # number of vllm replicas 
+  batch_size: 128 # ray pipeline internal batch size (used for map_batches call internally). Should usually be set to a value in [64, 128, 256] for best performance.
```
Technically with vlllm's batching we can do `batch_size = total_dataset_size / num_replicas` right? Is there a reason we don't do this in rayllm batch? I realize there's some continuous batching that rayllm_batch provides, but theoretically it's still better to just send everything and let vllm's scheduler handle the rest?

erictang000 replied:
yeah i think that makes sense to me - already added code to do this for the total_dataset_size < batch_size case, but yeah we could probably just do it for everything...

SumanthRH replied:
SG. oh wait, let me port over the commens to the main PR for consistency

erictang000 committed:
address some comments (add response object and add separate inference + ray functions)

erictang000 committed:
small comment use_ray

erictang000 committed:
resolve some more comments

erictang000 committed:
reduce workload

erictang000 committed:
changes

erictang000 committed:
fix ProcessPoolExecutor + response ray sigsev bug

erictang000 committed:
Merge branch 'main' of https://github.com/erictang000/SkyThought into HEAD

erictang000 committed:
add comment

SumanthRH committed:
Refactor model-specific configs and move data curation scripts (#60)

Another refactor PR on top of #23 now focused on model-specific configurations and data generation.

- Model-specific system prompts, user templates etc are best left to be in the a YAML file. 
- TaskHandler should be model agnostic, since we want to have a consistent evaluation logic for all tasks
- Data curation scripts for different Sky-T1 models should live outside the `skythought_evals` package. These are mostly scripts focused on a particular data curation task like filtering, rewriting etc. My proposal is to place common scripts in `scripts/ `. A guide for obtaining the final training data + training commands for different Sky-T1 models should be placed in `recipes/` . For now, all data curation scripts are in the `scripts` folder . 
- Adds a new `system-prompt-template` CLI flag. User can leverage available templates like those for sky-T1, Qwen, etc for a different model during evaluation.

SumanthRH committed:
Merge branch 'main' of https://github.com/NovaSky-AI/SkyThought into rayllm

Signed-off-by: SumanthRH <sumanthrh@anyscale.com>
